content/post/language_migrations.md

```diff
@@ -0,0 +1,512 @@
+\\end{equation}
+\\begin{equation}\\label{emd}
+\\sum_ {i=1}^N \\sum_ {j=1}^N x_ {ij} = 1
+\\end{equation}
+
+Let's make a little aside here. First, note that for an unbiased matter, all edges' "costs" are positive and set equal to one. Furthermore, to make our problem stick to classical [flow minimization formulation](https://en.wikipedia.org/wiki/Minimum-cost_flow_problem), we would have to add an artificial source and sink on both sides of our [bipartite graph](https://en.wikipedia.org/wiki/Bipartite_graph) to ensure flow conservation. However, it doesn't change anything, especially not linearity of the problem. If you want a brief insight about such a transformation, you can take look at the last slides of [Stanford's CS97SI](https://web.stanford.edu/class/cs97si/08-network-flow-problems.pdf).
+
+## Transition Matrix
+
+Here the main code to compute the transition matrix of a specific GitHub user and between two corresponding consecutive years. Note that the main function we are using which is provided by the [`PyEMD`](https://pypi.python.org/pypi/pyemd) package is [`emd_with_flow`](https://pypi.python.org/pypi/pyemd).
+
+```Python
+def get_transition_matrix_emd(user, year):
+
+    # lang2keep is the list of the 22 programming languages we kept
+    # stairvalue() is the step function that comes from quantization
+
+    # Build user's reactors for year and year+1
+    reactors = zeros((len(lang2keep),2), order='f')
+
+    for ind, code_in_lang in enumerate(dataset[user]):
+        lang = code_in_lang[0]
+
+        if lang in lang2keep:
+            for y, qtt_coded_year in code_in_lang[1]:
+                if y == year:
+                    reactors[lang2keep.index(lang),0] = stairvalue(qtt_coded_year)
+
+                elif y == year+1:
+                    reactors[lang2keep.index(lang),1] = stairvalue(qtt_coded_year)
+
+    if (sum(reactors[:,0]) == 0) or (sum(reactors[:,1]) == 0):
+        # no transition to consider
+        P = zeros((len(lang2keep), len(lang2keep)))
+        return P
+
+    else:
+        # Normalization of reactors
+        for i in [0, 1]:
+            reactors[:,i] = [j/sum(reactors[:,i]) for j in reactors[:,i]]
+
+    # compute the Earth Mover's distance between the 2 reactors thanks to the emd_with_flow function
+    distance = np.ones((len(lang2keep), len(lang2keep)))
+    dist, P = emd_with_flow(reactors[:,0], reactors[:,1], distance)
+    P = np.asarray(P)
+
+    return P
+```
+<p align="center" class="caption">Function to compute a transition matrix in Python</p>
+
+
+Finally, after summing the flow matrices over all GitHub users, and over the last 16 years (we will talk this last assumption later in this post), we obtain our own transition matrix. Let's now compare it to the contingency table Erik got from Google queries. To get insights about how these nice matrices have been plotted, I invite you to take a look at [Erik's script](https://github.com/erikbern/eigenstuff/blob/master/analyze.py).
```

Comment:
we will consider the latter assumption below

we obtain the resulting transition matrix

Erik compiled